### PR TITLE
Heap up metric when failed to post, and attach it next time.

### PR DESCRIPTION
**Pull request by Krout0n**
When agent fails to post metric, it should heap up for the next time. I implement that and add a test.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -107,3 +111,61 @@ pub mod host_meta;
...
+            Self {
+                client,
+                config: config::Config::default(),
+                host_id: "host_id_1".to_string(),
+                metric: vec![],
+            }
+        }
+    }
+
+    #[tokio::test]
```
Comment (Krout0n) [inline src/lib.rs:135]:
```rust
#[test]
async fn test_name() {
 ...
}
```
async functions cannot be used for tests. So, we should use `#[tokio::test]`

File: Cargo.toml [hunk 2]
```diff
@@ -24,4 +24,7 @@ lto = true
 opt-level = "s"
 
 [dev-dependencies]
+error-chain = "0.12.2"
+futures = "0.3"
 mockall = "0.8.0"
```
Comment (Krout0n) [inline Cargo.toml:29]: These libs are used in mackerel_client. In production code, they ain't used **directly**. So, I add them in dev-dependencies.

File: src/lib.rs [hunk 3]
```diff
@@ -91,13 +95,13 @@ impl Agent {
...
-        if result.is_err() {
-            dbg!(result.err());
-        }
+    async fn send_metric(&mut self, val: MetricValue) {
+        let mut metric: Vec<_> = HostMetricWrapper(&self.host_id, val).into();
+        metric.extend(self.metric.clone());
+        let result = self.client.post_metrics(metric.clone()).await;
+        // If Ok, then heaped metric must be empty, else extend it.
+        // TODO: Drop metric if it is too ancient.
+        self.metric = if result.is_ok() { vec![] } else { metric };
```
Comment (ne-sachirou) [inline src/lib.rs:104]: Instead of silently fails, It's better to log a warning here.

**Review by ne-sachirou — APPROVED** [review]
🙆‍♀️ great

File: src/lib.rs (hunk 3 above)
Comment (Krout0n) [reply src/lib.rs:104]: Yes, we need logger....:innocent:

**Comment by Krout0n** [comment]
Thank you!

File: src/lib.rs [hunk 4]
```diff
@@ -91,13 +95,13 @@ impl Agent {
...
-    async fn send_metric(&self, val: MetricValue) {
-        let metric = HostMetricWrapper(&self.host_id, val).into();
-        // TODO: error handling.
-        let result = self.client.post_metrics(metric).await;
-        if result.is_err() {
-            dbg!(result.err());
-        }
+    async fn send_metric(&mut self, val: MetricValue) {
+        let mut metric: Vec<_> = HostMetricWrapper(&self.host_id, val).into();
+        metric.extend(self.metric.clone());
```
Comment (itchyny) [inline src/lib.rs:100]: It should limit the size of metrics, otherwise it stops after reaching the maximum body size.

Comment (itchyny) [reply src/lib.rs:100]: For reference, metrics sender of container agent keeps the metrics for 6 hours and sends metrics in 3 minutes at most.
https://github.com/mackerelio/mackerel-container-agent/blob/36cf0bb103d1645fb733c214fe142207b1318838/metric/sender.go#L22-L50

Comment (Krout0n) [reply src/lib.rs:100]: Exactly. After reaching, should Agent drop a metric from the most previous for heaping new one?

Comment (Krout0n) [reply src/lib.rs:100]: Thank you for your comment! If you are still interested in, please review #23 too.

**Pull request merged**